homeassistant/components/tplink/common.py

```diff
@@ -0,0 +1,189 @@
+    """Get devices through discovery."""
+    _LOGGER.debug("Discovering devices")
+    devices = await async_has_discoverable_devices(hass)
+    _LOGGER.info(
+        "Discovered %s TP-Link smart home device(s)",
+        len(devices)
+    )
+
+    lights = []
+    switches = []
+
+    for dev in devices.values():
+        if isinstance(dev, SmartPlug):
+            try:
+                if dev.is_dimmable:  # Dimmers act as lights
+                    lights.append(dev)
+                else:
+                    switches.append(dev)
+            except SmartDeviceException as ex:
+                _LOGGER.error("Unable to connect to device %s: %s",
+                              dev.host, ex)
+
+        elif isinstance(dev, SmartBulb):
+            lights.append(dev)
+        else:
+            _LOGGER.error("Unknown smart device type: %s", type(dev))
+
+    return SmartDevices(lights, switches)
+
+
+def async_get_static_devices(config_data) -> SmartDevices:
+    """Get statically defined devices in the config."""
+    _LOGGER.debug("Getting static devices")
+    lights = []
+    switches = []
+
+    for type_ in [CONF_LIGHT, CONF_SWITCH]:
+        for entry in config_data[type_]:
+            host = entry['host']
+
+            if type_ == CONF_LIGHT:
+                lights.append(SmartBulb(host))
+            elif type_ == CONF_SWITCH:
+                switches.append(SmartPlug(host))
+
+    return SmartDevices(
+        lights,
+        switches
+    )
+
+
+def async_add_entities_retry(
+        hass: HomeAssistantType,
+        async_add_entities: Callable[[List[Any], bool], None],
+        objects: List[Any],
+        callback: Callable[[Any, Callable], None],
+        interval: timedelta = timedelta(seconds=30)
+):
+    """
+    Add entities now and retry later if issues are encountered.
+
+    If the callback throws an exception or returns false, that
+    object will try again a while later.
+    This is useful for devices that are not online when hass starts.
+    :param hass:
+    :param async_add_entities: The callback provided to a
+    platform's async_setup.
+    :param objects: The objects to create as entities.
+    :param callback: The callback that will perform the add.
+    :param interval: THe time between attempts to add.
+    :return: A callback to cancel the retries.
+    """
+    add_objects = objects.copy()
+
+    def dummy_cancel():
+        pass
+
+    def process_objects(*args):
+        success_indexes = []
+
+        # Process each object.
+        for i, add_object in enumerate(add_objects):
+            # Call the individual item callback.
+            try:
+                _LOGGER.debug(
+                    "Attempting to add object of type %s",
+                    type(add_object)
+                )
+                result = callback(add_object, async_add_entities)
+            except Exception as ex:  # pylint: disable=broad-except
```

Comment:
Please catch the exception that you expect to be raised when something goes wrong.